### PR TITLE
[AMDGPU] Constrain use LiveMask by the operand's LaneMask for RP calculation.

### PR DESCRIPTION
**Pull request by jrbyrnes**
For speculative RP queries, recede may calculate inaccurate masks for subreg uses. Previously, the calculation would look at any live lane for the use at the position of the MI in the LIS. This also adds lanes for any subregs which are live at but not used by the instruction. By constraining against the getSubRegIndexLaneMask for the operand's subreg, we are sure to not pick up on these extra lanes.

For current clients of recede, this is not an issue. This is because 1. the current clients do not violate the program order in the LIS, and 2. the change to RP is based on the difference between previous mask and new mask. Since current clients are not exposed to this issue, this patch is sort of NFC.

Co-authored-by: Valery Pykhtin <Valery.Pykhtin@amd.com>

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h [hunk 1]
```diff
@@ -265,13 +265,14 @@ class GCNDownwardRPTracker : public GCNRPTracker {
                const LiveRegSet *LiveRegsCopy = nullptr);
 };
 
-LaneBitmask getLiveLaneMask(unsigned Reg,
-                            SlotIndex SI,
+LaneBitmask getLiveLaneMask(unsigned Reg, SlotIndex SI,
                             const LiveIntervals &LIS,
-                            const MachineRegisterInfo &MRI);
+                            const MachineRegisterInfo &MRI,
+                            LaneBitmask Mask = LaneBitmask::getAll());
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.h:271]: Parameter name could be more specific, and should be documented.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 2]
```diff
@@ -259,33 +259,40 @@ static void
...
+    });
+
+    auto &P = I == RegMaskPairs.end()
+                  ? RegMaskPairs.emplace_back(Reg, LaneBitmask::getNone())
+                  : *I;
 
-    LaneBitmask UseMask;
-    auto &LI = LIS.getInterval(Reg);
+    P.LaneMask |= MO.getSubReg() ? TRI.getSubRegIndexLaneMask(MO.getSubReg())
+                                 : MRI.getMaxLaneMaskForVReg(Reg);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:280]: I believe TRI.getSubRegIndexLaneMask handles the 0 case just fine

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:280]: `MRI.getMaxLaneMaskForVReg` uses the `TRC` for width-aware `LaneMask`s whereas `TRI.getSubRegIndexLaneMask just returns `~0u`

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h [hunk 3]
```diff
@@ -265,14 +265,18 @@ class GCNDownwardRPTracker : public GCNRPTracker {
                const LiveRegSet *LiveRegsCopy = nullptr);
 };
 
+/// \returns the LaneMask of live lanes of \p Reg at position \p SI. Only the
+/// active lanes of \p MaxLaneMask will be set in the return value. This is
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.h:269]: Seems more like a filter than MaxLaneMask?

**Review by arsenm — APPROVED** [review]

**Comment by jrbyrnes** [comment]
Thanks for review

https://github.com/llvm/llvm-project/commit/5cb6b15568c1554a2ea89fee849ed0e27b785ee0

**Pull request closed without merging**